File: bot/cryptopay.py
"""Минимальный асинхронный клиент Crypto Pay API (@CryptoBot).

Документация: https://help.crypt.bot/crypto-pay-api
"""

from typing import Any

import aiohttp

MAINNET_URL = "https://pay.crypt.bot/api"
TESTNET_URL = "https://testnet-pay.crypt.bot/api"


class CryptoPayError(Exception):
    def __init__(self, code: int, name: str):
        self.code = code
        self.name = name
        super().__init__(f"CryptoPay error {code}: {name}")


class CryptoPay:
    def __init__(self, token: str, testnet: bool = False):
        self._token = token
        self._base_url = TESTNET_URL if testnet else MAINNET_URL
        self._session: aiohttp.ClientSession | None = None

    async def _request(self, method: str, **params: Any) -> Any:
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                headers={"Crypto-Pay-API-Token": self._token}
            )
        payload = {k: v for k, v in params.items() if v is not None}
        async with self._session.post(
            f"{self._base_url}/{method}", json=payload
        ) as resp:
            data = await resp.json()
        if not data.get("ok"):
            err = data.get("error", {})
            raise CryptoPayError(err.get("code", 0), err.get("name", "UNKNOWN"))
        return data["result"]

    async def close(self) -> None:
        if self._session and not self._session.closed:
            await self._session.close()

    async def get_me(self) -> dict:
        return await self._request("getMe")

    async def create_invoice(
        self,
        asset: str,
        amount: float,
        description: str | None = None,
        payload: str | None = None,
        expires_in: int | None = None,
    ) -> dict:
        return await self._request(
            "createInvoice",
            currency_type="crypto",
            asset=asset,
            amount=str(amount),
            description=description,
            payload=payload,
            expires_in=expires_in,
        )

    async def get_invoices(self, invoice_ids: list[int]) -> list[dict]:
        result = await self._request(
            "getInvoices", invoice_ids=",".join(map(str, invoice_ids))
        )
        return result.get("items", [])

    async def transfer(
        self,
        user_id: int,
        asset: str,
        amount: float,
        spend_id: str,
        comment: str | None = None,
    ) -> dict:
        """Выплата пользователю. spend_id — ключ идемпотентности."""
        return await self._request(
            "transfer",
            user_id=user_id,
            asset=asset,
            amount=str(amount),
            spend_id=spend_id,
            comment=comment,
            disable_send_notification=False,
        )

    async def get_balance(self) -> list[dict]:
        return await self._request("getBalance")

File: worker/src/cryptopay.ts
// Минимальный клиент Crypto Pay API (@CryptoBot) поверх fetch.
// Документация: https://help.crypt.bot/crypto-pay-api

const MAINNET_URL = "https://pay.crypt.bot/api";
const TESTNET_URL = "https://testnet-pay.crypt.bot/api";

export class CryptoPayError extends Error {
  constructor(
    public code: number,
    public errorName: string,
  ) {
    super(`CryptoPay error ${code}: ${errorName}`);
  }
}

export interface Invoice {
  invoice_id: number;
  status: string; // active | paid | expired
  asset: string;
  amount: string;
  bot_invoice_url: string;
  payload?: string;
}

export class CryptoPay {
  private baseUrl: string;

  constructor(
    private token: string,
    testnet = false,
  ) {
    this.baseUrl = testnet ? TESTNET_URL : MAINNET_URL;
  }

  private async request<T = unknown>(method: string, params: Record<string, unknown> = {}): Promise<T> {
    const payload: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(params)) {
      if (v !== undefined && v !== null) payload[k] = v;
    }
    const resp = await fetch(`${this.baseUrl}/${method}`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        "Crypto-Pay-API-Token": this.token,
      },
      body: JSON.stringify(payload),
    });
    const data = (await resp.json()) as {
      ok: boolean;
      result?: T;
      error?: { code?: number; name?: string };
    };
    if (!data.ok) {
      throw new CryptoPayError(data.error?.code ?? 0, data.error?.name ?? "UNKNOWN");
    }
    return data.result as T;
  }

  getMe() {
    return this.request<{ name?: string }>("getMe");
  }

  createInvoice(opts: {
    asset: string;
    amount: number;
    description?: string;
    payload?: string;
    expires_in?: number;
  }): Promise<Invoice> {
    return this.request<Invoice>("createInvoice", {
      currency_type: "crypto",
      asset: opts.asset,
      amount: String(opts.amount),
      description: opts.description,
      payload: opts.payload,
      expires_in: opts.expires_in,
    });
  }

  async getInvoices(invoiceIds: number[]): Promise<Invoice[]> {
    const result = await this.request<{ items?: Invoice[] }>("getInvoices", {
      invoice_ids: invoiceIds.join(","),
    });
    return result.items ?? [];
  }

  /** Выплата пользователю. spendId — ключ идемпотентности. */
  transfer(opts: {
    user_id: number;
    asset: string;
    amount: number;
    spend_id: string;
    comment?: string;
  }) {
    return this.request("transfer", {
      user_id: opts.user_id,
      asset: opts.asset,
      amount: String(opts.amount),
      spend_id: opts.spend_id,
      comment: opts.comment,
      disable_send_notification: false,
    });
  }

  getBalance() {
    return this.request<{ currency_code: string; available: string }[]>("getBalance");
  }
}

/**
 * Проверка подписи вебхука Crypto Pay:
 * заголовок crypto-pay-api-signature = HMAC-SHA256(body), ключ — SHA256(token).
 */
export async function checkCryptoPaySignature(
  token: string,
  body: string,
  signature: string | null,
): Promise<boolean> {
  if (!signature) return false;
  const enc = new TextEncoder();
  const secret = await crypto.subtle.digest("SHA-256", enc.encode(token));
  const key = await crypto.subtle.importKey(
    "raw",
    secret,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const mac = await crypto.subtle.sign("HMAC", key, enc.encode(body));
  const hex = [...new Uint8Array(mac)].map((b) => b.toString(16).padStart(2, "0")).join("");
  return hex === signature.toLowerCase();
}
